### ISSUE TITLE
redux-auth-wrapper Version 2 Proposal

### ISSUE THREAD

**Issue by mjrussell** [mjrussell/redux-auth-wrapper#145]
Ok I've got an initial version of version 2 of redux-auth-wrapper and I'd like some initial feedback before I push on updating the tests and documentation. Overall, I think its is much more flexible than the v1 version although a bit more complicated to use. I'd like to provide some more convenience imports for users trying to get something simple working fast but let more advanced users use just the pieces they want.

For the initial reasons behind the rewrite, please see #133. I've been considering moving towards this direction for a little while now, but React Router v4 support sort of pushed it sooner than I anticipated.

The current overall changes are the following (can skip to the bottom for the examples):

### authWrapper focuses on rendering
The `authWrapper` (no more `AuthWrapper`, its an HOC not a Component) is completely separated from any knowledge about routing or redirection. This makes it less awkward to use when hiding components or showing an alternate component when the predicate fails.

The `authWrapper` also no longer includes `connect` by default, exporting an unconnected version by default (more on the connected version later)

The parameters to the config object of `authWrapper` are:

* `AuthenticatingComponent` - Renamed from LoadingComponent but otherwise the same
* `FailureComponent` - No Change
* `wrapperDisplayName` - No Change
* `predicate` - No Change
* `propMapper` - The default has changed to not exclude any props, but due to other changes should be essentially the same

This returns a component which expects 2 props:
* `authData` - No change
* `isAuthenticating` - No chance (optional, defaults to false)

Since most users (myself included) use this with redux, there is a connectedAuthWrapper that adds `authSelector` and `authenticatingSelector` from the previous version. The result is that the import from `redux-auth-wrapper/connectedAuthWrapper` looks much like the original without some of redirection params.
```js

export const userIsAuthenticated = authWrapper({
  authSelector: state => state.user.data,
  authenticatingSelector: state => state.user.isLoading,
  AuthenticatingComponent: Loading,
  FailureComponent: AuthFailureRedirect,
  wrapperDisplayName: 'UserIsAuthenticated'
})

export const userIsAdmin = authWrapper({
  authSelector: state => state.user.data,
  FailureComponent: AdminFailureRedirect,
  predicate: user => user.isAdmin,
  wrapperDisplayName: 'UserIsAdmin'
})

export const userIsNotAuthenticated = authWrapper({
  authSelector: state => state.user,
  wrapperDisplayName: 'UserIsNotAuthenticated',
  // Want to redirect the user when they are done loading and authenticated
  predicate: user => user.data === null && user.isLoading === false,
  FailureComponent: LoggedInRedirect
})
```

The next section will cover what the `FailureComponent` looks like.

### FailureComponent

For displaying an alternate component with the wrapper, you can use `FailureComponent` like in the previous version. Now, if you want to perform a redirect, it you pass a `FailureComponent` that performs the redirect. You don't have to use the ones from redux-auth-wrapper but there are helpers for building, parsing and performing the redirect.

The redirect component used is dead simple, its just the following:
```js
export default class Redirect extends Component {

  static propTypes = {
    redirectPath: PropTypes.string.isRequired,
    redirect: PropTypes.func.isRequired
  };

  componentWillMount() {
    this.props.redirect(this.props, this.props.redirectPath)
  }

  render() {
    // Redirect should happen before this is rendered
    return null
  }
}
```
So the other helpers are for passing down the `redirect` function and the `redirectPath`.

### Redirection

There are currently two redirect helpers (for history/react-router v3 and history/react-router v4) exported by redux-auth-wrapper, they both take a config object with the following defaults:

* `redirectQueryParamName` - Same as before
* `locationSelector(props) : Location` - selector from props to a location object

and returns an object with the following:

* `getRedirect(props): String` - Returns the redirectQuery as a path, returns null if no query present
* `createRedirect(allowRedirectBack): (props, redirectPath): Location` - Curried function that takes the `allowRedirect` param, and returns a new function that can be used by the `Redirect` component to create a location for the redirect

### Examples

So the only tricky part is how to pass the redirection through to the `Redirect` component. Its a little awkward, but gives us a solution that works for react-router3, react-router-redux4, and react-router4:

#### react-router-redux4 (react-router 3)

I use connect here to link up the `redirect` with the `replace` from react-router-redux:
```js
import { connect } from 'react-redux'
import redirectHelperBuilder from 'redux-auth-wrapper/history3'
import Redirect from 'redux-auth-wrapper/redirect'
import authWrapper from 'redux-auth-wrapper/connectedAuthWrapper'
import { routerActions } from 'react-router-redux'

import { Loading } from './components'

const redirectHelper = redirectHelperBuilder({})

const createRedirect = allowRedirectBack => dispatch => ({
  redirect: (...args) => {
    const redirectLoc = redirectHelper.createRedirect(allowRedirectBack)(...args)
    dispatch(routerActions.replace(redirectLoc))
  }
})

const AuthFailureRedirect = connect(() => ({
  redirectPath: '/login'
}), createRedirect(true))(Redirect)

export const userIsAuthenticated = authWrapper({
  authSelector: state => state.user.data,
  authenticatingSelector: state => state.user.isLoading,
  AuthenticatingComponent: Loading,
  FailureComponent: AuthFailureRedirect,
  wrapperDisplayName: 'UserIsAuthenticated'
})

const AdminFailureRedirect = connect(() => ({
  redirectPath: '/'
}), createRedirect(false))(Redirect)

export const userIsAdmin = authWrapper({
  authSelector: state => state.user.data,
  FailureComponent: AdminFailureRedirect,
  predicate: user => user.isAdmin,
  wrapperDisplayName: 'UserIsAdmin'
})

const LoggedInRedirect = connect((state, ownProps) => ({
  redirectPath: redirectHelper.getRedirect(ownProps) || '/foo'
}), createRedirect(false))(Redirect)

export const userIsNotAuthenticated = authWrapper({
  authSelector: state => state.user,
  wrapperDisplayName: 'UserIsNotAuthenticated',
  // Want to redirect the user when they are done loading and authenticated
  predicate: user => user.data === null && user.isLoading === false,
  FailureComponent: LoggedInRedirect
})
```

#### react-router 4 (no react-router-redux)

Here, we dont actually need to connect the `Redirect`, so I use [https://github.com/acdlite/recompose/blob/master/docs/API.md#withprops](withProps) from recompose to add properties. I could have also just made a function component that adds the properties directly.
```js
import redirectHelperBuilder from 'redux-auth-wrapper/history4'
import Redirect from 'redux-auth-wrapper/redirect'
import authWrapper from 'redux-auth-wrapper/connectedAuthWrapper'
import { withProps } from 'recompose'

import Loading from './components/Loading'

const redirectHelper = redirectHelperBuilder({})

const createRedirect = (allowRedirectBack, history) => (...args) => {
  const redirectLoc = redirectHelper.createRedirect(allowRedirectBack)(...args)
  history.replace(redirectLoc)
}

const AuthFailureRedirect = withProps((props) => ({
  redirectPath: '/login',
  redirect: createRedirect(true, props.history)
}))(Redirect)

export const userIsAuthenticated = authWrapper({
  authSelector: state => state.user.data,
  authenticatingSelector: state => state.user.isLoading,
  AuthenticatingComponent: Loading,
  FailureComponent: AuthFailureRedirect,
  wrapperDisplayName: 'UserIsAuthenticated'
})

const AdminFailureRedirect = withProps((props) => ({
  redirectPath: '/',
  redirect: createRedirect(false, props.history)
}))(Redirect)

export const userIsAdmin = authWrapper({
  authSelector: state => state.user.data,
  FailureComponent: AdminFailureRedirect,
  predicate: user => user.isAdmin,
  wrapperDisplayName: 'UserIsAdmin'
})

const LoggedInRedirect = withProps((props) => ({
  redirectPath: redirectHelper.getRedirect(props) || '/foo',
  redirect: createRedirect(false, props.history)
}))(Redirect)

export const userIsNotAuthenticated = authWrapper({
  authSelector: state => state.user,
  wrapperDisplayName: 'UserIsNotAuthenticated',
  // Want to redirect the user when they are done loading and authenticated
  predicate: user => user.data === null && user.isLoading === false,
  FailureComponent: LoggedInRedirect
})
```

Feel free to check out or run the examples on the branch: https://github.com/mjrussell/redux-auth-wrapper/tree/rewrite-v2. Both the `loading` and `loading-rrv4` are updated.

As you can see, its largely the same behavior, API, with much more flexibility. The only awkward part is with providing the props to the `Redirect` component. I could provide helpers to users to make this a little easier, for connected and unconnected variants and then let advanced users wire it up themselves.

**Comment by oyeanuj** [comment]
Hey @mjrussell, this looks good - thanks for the detailed post. One thing that came to my mind was SSR support. In #133, you mentioned some static method for that.. but wondering if SSR API will change in this version (I assume yes, since there is no `onEnter` in RR4?)

**Comment by mjrussell** [comment]
@oyeanuj Yup, it would definitely change. Maybe the history3 helper can export an onenter method. In this case I dont think it would need to be static on the auth-wrapper since it doesnt even know about routing anymore

**Comment by dezoxel** [comment]
@mjrussell I'm trying to use it in my project using `npm i git://github.com/mjrussell/redux-auth-wrapper#rewrite-v2 --save`.

I've found that entry point should be `lib/index.js`, but after running `npm run build` there is no index.js file created. How can I get `lib/index.js`? How you do it before publishing to npm?

Maybe you propose me another approach to try it out in read world environment?

**Comment by shai32** [comment]
@mjrussell I have some questions about the implementation:
1.Do you think this implementation is good, performance wise?
implementing a simple redirect as a component sounds like a big overhead,
Each route adds to 3 dummy components: The HOC, AuthenticatingComponent, FailureComponent
Maybe we need to also support redirectAction without creating a dummy component that wraps it.
To avoid the AuthenticatingComponent, we can  add a  AuthenticatingString, just to show a simple 'Loadding ..' for s simple use case.  

2. why do I need both  authSelector and predicate, can we just have predicate that get the state and evaluates if its true or false, eg:
```
const VisibleOnlyAdmin = UserAuthWrapper({
  predicate: state => state.user.isAdmin
})
```
just for simplicity, less definitions.
 
3. whats the default value for FailureComponent? is it null?

**Comment by mjrussell** [comment]
@dezoxel how are you trying to import it? I haven't fully decided how/if I want to group the imports into an "index". Probably at least the HOC, but I think its a good idea to keep the redirect helpers in their own files to allow for only importing the history version that is applicable.

In v1 I ran a prepbulish (basically npm build) before pushing to npm. You should be able to do the following in the short term for testing until I put something on npm:
````
npm i git://github.com/mjrussell/redux-auth-wrapper#rewrite-v2 --save
cd node_modules/redux-auth-wrapper
npm install
```
then in your source:
`import authWrapper from 'redux-auth-wrapper/lib/connectedAuthWrapper'

**Comment by mjrussell** [comment]
@shai32 

1. Creating a component especially one that will probably never mount (due to the redirection in componentWillMount) is a very tiny overhead, I find it hard to believe that it would cause a measurable performance change. I dont think changing AuthenticatingComponent to AuthenticatingString is a solution, what if you wanted to display a spinner or something else? Using a function (stateless) component should really limit any overhead found here.

2. The `authData` vs `predicate` is used to pass through the data down to sub components. If you use the default check (not null/empty) then you can be sure that the prop `authData` is not null/empty inside your wrapper component. This was absolutely necessary before major changes to react-redux in [v5](https://github.com/reactjs/react-redux/releases/tag/v5.0.0) because if you re-connected the `authDataSelector` lower in the component tree you did not have a guarantee that it would be rendered with the `authData` as not null/empty. This is because children could react to redux-store changes before their parents. I think with the new version of react-redux this is less useful now, except to avoid a subsequent connect for just the authData.

3. Yes, the default is `undefined` and it will render `null` in this [case](https://github.com/mjrussell/redux-auth-wrapper/blob/rewrite-v2/src/authWrapper.js#L48)

**Comment by shai32** [comment]
@mjrussell 
Still, adding component just to do redirect doesn't feel right. are you doing it only to support both v2 and v3?
 I meant AuthenticatingString alongside  AuthenticatingComponent, so we can chose one of the options.

do we need to check if AuthenticatingComponent is null? is it failing  if I don't define any AuthenticatingComponent?
![image](https://cloud.githubusercontent.com/assets/16700996/25170140/280cc36c-24f2-11e7-85bb-f898db2ec295.png)

**Comment by dezoxel** [comment]
@mjrussell I've tried new version, everything works as expected according to your examples. I was able to implement everything I need in terms of authentication and authorization. Overall I like it a lot.

**Comment by dezoxel** [comment]
I have a couple notes about API:

1. `authWrapper()` could be renamed to `withAuth()`
2. Add option like `onFailure`: 
```onFailure: (dispatch, redirectPath) => dispatch(goToLogin(redirectPath))```

Having FailureComponent is a nice feature if someone needs to render login failure screen, but it looks too complicated when implementing simple redirect. I would happy to have ability to just dispatch redux action with appropriate data.

**Comment by clintjansen** [comment]
Any updates on this? I would really like to use it with RRv4, thanks!

**Comment by mjrussell** [comment]
Sorry for the delay everyone, I know you are looking forward to using this with RRv4. I really appreciate the feedback provided so far, I hope to respond in turn to some of the comments once I have a little bit more free time. I've incorporated a bunch of it into an update that I just pushed tonight and published under the beta tag.

You can install it with:
`npm install redux-auth-wrapper@2.0-beta` 

I just pushed to npm the beta version. I think this API is mostly stabilized, but once again would appreciate any more feedback.

Essentially the changes I made don't prevent developers from using the code as above, but I've made the simple case of doing a simple redirect much easier. Here are the updated examples:

#### react-router-redux4 (react-router 3)
```js
import locationHelperBuilder from 'redux-auth-wrapper/lib/history3/locationHelper'
import { connectedReduxRedirect } from 'redux-auth-wrapper/lib/history3/redirect'
import { routerActions } from 'react-router-redux'

import { Loading } from './components'

const locationHelper = locationHelperBuilder({})

export const userIsAuthenticated = connectedReduxRedirect({
  redirectPath: '/login',
  authSelector: state => state.user.data,
  authenticatingSelector: state => state.user.isLoading,
  AuthenticatingComponent: Loading,
  redirectAction: routerActions.replace,
  wrapperDisplayName: 'UserIsAuthenticated'
})

export const userIsAdmin = connectedReduxRedirect({
  redirectPath: '/',
  allowRedirectBack: false,
  authSelector: state => state.user.data,
  predicate: user => user.isAdmin,
  redirectAction: routerActions.replace,
  wrapperDisplayName: 'UserIsAdmin'
})

export const userIsNotAuthenticated = connectedReduxRedirect({
  redirectPath: (state, ownProps) => locationHelper.getRedirectQuery(ownProps) || '/foo',
  allowRedirectBack: false,
  authSelector: state => state.user,
  // Want to redirect the user when they are done loading and authenticated
  predicate: user => user.data === null && user.isLoading === false,
  redirectAction: routerActions.replace,
  wrapperDisplayName: 'UserIsNotAuthenticated'
})
```

#### react router 4
```js
import locationHelperBuilder from 'redux-auth-wrapper/lib/history4/locationHelper'
import { connectedRouterRedirect } from 'redux-auth-wrapper/lib/history4/redirect'

import Loading from './components/Loading'

const locationHelper = locationHelperBuilder({})

export const userIsAuthenticated = connectedRouterRedirect({
  redirectPath: '/login',
  authSelector: state => state.user.data,
  authenticatingSelector: state => state.user.isLoading,
  AuthenticatingComponent: Loading,
  wrapperDisplayName: 'UserIsAuthenticated'
})

export const userIsAdmin = connectedRouterRedirect({
  redirectPath: '/',
  allowRedirectBack: false,
  authSelector: state => state.user.data,
  predicate: user => user.isAdmin,
  wrapperDisplayName: 'UserIsAdmin'
})

export const userIsNotAuthenticated = connectedRouterRedirect({
  redirectPath: (state, ownProps) => locationHelper.getRedirectQuery(ownProps) || '/foo',
  allowRedirectBack: false,
  authSelector: state => state.user,
  // Want to redirect the user when they are done loading and authenticated
  predicate: user => user.data === null && user.isLoading === false,
  wrapperDisplayName: 'UserIsNotAuthenticated'
})
```

Basically the `'redux-auth-wrapper/lib/historyX/redirect` each export two versions:
`connectedRouterRedirect` and `connectedReduxRedirect`. The `connectedRouterRedirect` will use the the history/router directly to perform the redirection while the `connectedReduxRedirect` will use the provided `redirectAction`

**Comment by mjrussell** [comment]
I think I may have broken support for the `connectedRouterRedirect` for react router 3, I'll try to fix that tomorrow and push an update. And I'll start working the tests so its easier to catch these things and get that coverage back up!

**Comment by mjrussell** [comment]
@shai32 @dezoxel Does the new version help address some of the issues you found? At least the concerns for getting the simple redirection going?

**Comment by mjrussell** [comment]
I saw the 👍 but wasn't sure if that was just encouragement for me getting it published to npm 😉

**Comment by shai32** [comment]
@mjrussell  the redirectAction works perfect!! , thats exactly what I needed. 

regarding the "AuthenticatingString" I asked for, I am not sure it is a good idea. 
I think that you are right, the AuthenticatingComponent is better.

**Comment by piuccio** [comment]
I tried your `2.0.0-beta` branch on react-native and I get the error 

```
error: bundling: UnableToResolveError: Unable to resolve module `url`
```

same happens with version 1 as mentioned in #136

**Comment by piuccio** [comment]
One thing I noticed is that `redirectPath` is mandatory but gets ignored if you use a `FailureComponent`. If you remove it you'll get `redirectPath must be either a string or a function`

**Comment by prescottprue** [comment]
Great work on this @mjrussell, `react-redux-firebase` uses `redux-auth-wrapper` in its routing examples, so its awesome to see `react-router` 4 will be supported. Looking forward to updating them when this changes.

**Comment by mjrussell** [comment]
I had some time to work more on this tonight and I got every test but one passing for both react router 3 and react router 4 so that is good news. I also tweaked the internal API slightly so that you can pass the allowRedirectBack and  queryParamName to the connectedRouterRedirect and connectedReduxRedirect helpers. So those match even more closely now to the original API so it should be an easier transition for users.

I've published a new version to that same beta tag

Regarding React Native support, this library has always wanted to support RN but since we've never had any tests or examples it's easy to break and not well supported. I would definitely like to improve that going forward with V2. I think you should not be using the connected redirect helpers for react native (that's where the stringify was imported but didn't have a dep entry). I don't think react native's routers support the same location objects. Using the test matrix I've got set up now should allow for adding react native tests in the future and keeping it up to date.

**Comment by mjrussell** [comment]
As an FYI here are the remaining items I see before releasing v2 
- [ ] Get code coverage back up to 100%. This where it stands [now](https://coveralls.io/builds/11383784):
![image](https://cloud.githubusercontent.com/assets/1641895/25750833/d252bec2-3167-11e7-9552-a5b290c31d4f.png)

- [ ] All examples running (this might only include 1 each for rrv3 and rrv4)
- [ ] Update documentation

For the docs, since there is a lot more flexibility and ways to use the library, I was thinking about trying to port it to gitbooks which I've used before and hosting it with github pages. The README is getting a bit long 😉

**Comment by mjrussell** [comment]
Coverage is all set, at 100% for all statements, branches, functions and lines! 🎉 

There is also an `onEnter` helper exported for use with SSR. (@oyeanuj will be glad to hear that)

And those two examples are up to date. I think the API is pretty stable now. On to the docs

**Comment by mjrussell** [comment]
Thanks for all the help everyone and patience. I think Im basically there. Just working some minor documentation and making sure there are no issues with the final beta tag. Otherwise will plan on releasing 2.0 officially by the end of the week

**Comment by JulianJorgensen** [comment]
@mjrussell Really great work! In v1 I used authSelector to pass down auth data to my components. In v2 I'm a bit confused if this is still doable. In the docs you suggested to connect at a lower level. Ideally I'd like the data passed down before `componentWillMount()`.
Cheers

**Comment by mjrussell** [comment]
@JulianJorgensen thanks!

Originally when this library was created, react-redux couldn't guarantee that connected components below a parent that was connected got their state updates after the parent updated and possibly re-rendered. Since this is a guarantee now, and because the old `authSelector` made it difficult for users with boolean fields I ultimately removed it.

I would suggest using connect for your auth data with your old `authSelector` where ever you need it.  You can see in the examples, I just connect it to the components that need the authData - https://github.com/mjrussell/redux-auth-wrapper/blob/master/examples/react-router-3/components/Foo.js#L9

If you want your HOCs to still pass down the `authData` to every component its applied to, you can compose it with connect:
```js
import { compose } from redux
import { connect } from react-redux
import { connectedRouterRedirect } from 'redux-auth-wrapper/history3/redirect'

export const userIsAuthenticated = compose(
  connectedRouterRedirect({
    redirectPath: '/login',
    authenticatedSelector: state => state.user.data !== null,
    authenticatingSelector: state => state.user.isLoading,
    AuthenticatingComponent: Loading,
    wrapperDisplayName: 'UserIsAuthenticated'
  }),
  connect((state) => ({ authData: authSelector(state) }))
);
```

Or if you find the compose syntax difficult to read:
```js
const authRedirect =  connectedRouterRedirect({
  redirectPath: '/login',
  authenticatedSelector: state => state.user.data !== null,
  authenticatingSelector: state => state.user.isLoading,
  AuthenticatingComponent: Loading,
  wrapperDisplayName: 'UserIsAuthenticated'
});

export const userIsAuthenticated = (WrappedComponent) => authRedirect(connect((state) => ({ authData: authSelector(state) }))(WrappedComponent))
```

**Comment by prescottprue** [comment]
@mjrussell I agree, great work! Going to update the `react-redux-firebase` docs based on these updates 😄

**Comment by asCii88** [comment]
@prescottprue I would be really happy if you updated the docs for `react-redux-firebase` I'm struggling to get it to work myself :)
